# BÀI 1: KHÔI PHỤC MASKED TOKEN

In [4]:
from transformers import pipeline

mask_filler = pipeline("fill-mask", model="bert-base-uncased")
input_sentence = "Hanoi is the [MASK] of Vietnam."
predictions = mask_filler(input_sentence, top_k=5)

print(f"Câu gốc: {input_sentence}\n")
for i, pred in enumerate(predictions, 1):
    print(f"{i}. Dự đoán: '{pred['token_str']}' với độ tin cậy: {pred['score']:.4f}")
    print(f"   -> Câu hoàn chỉnh: {pred['sequence']}\n")

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`
d:\10. ky1nam4\NLP\venv\lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in E:\hf_cache\hub\models--bert-base-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/win

Câu gốc: Hanoi is the [MASK] of Vietnam.

1. Dự đoán: 'capital' với độ tin cậy: 0.9991
   -> Câu hoàn chỉnh: hanoi is the capital of vietnam.

2. Dự đoán: 'center' với độ tin cậy: 0.0001
   -> Câu hoàn chỉnh: hanoi is the center of vietnam.

3. Dự đoán: 'birthplace' với độ tin cậy: 0.0001
   -> Câu hoàn chỉnh: hanoi is the birthplace of vietnam.

4. Dự đoán: 'headquarters' với độ tin cậy: 0.0001
   -> Câu hoàn chỉnh: hanoi is the headquarters of vietnam.

5. Dự đoán: 'city' với độ tin cậy: 0.0001
   -> Câu hoàn chỉnh: hanoi is the city of vietnam.



1. Mô hình đã dự đoán đúng từ "capital" không?

   Có, mô hình đã dự đoán đúng từ "capital" với độ tin cậy rất cao 
   (99%). Điều này cho thấy mô hình BERT đã học được mối quan hệ 
   ngữ nghĩa giữa "Hanoi" và "capital of Vietnam" rất tốt.

2. Tại sao các mô hình Encoder-only như BERT lại phù hợp cho tác vụ này?
   - BERT sử dụng cơ chế Self-Attention hai chiều (bidirectional), cho phép mô hình 
     xem xét cả các từ đứng trước và sau token [MASK].
   - Trong câu "Hanoi is the [MASK] of Vietnam", BERT có thể nhìn thấy cả "Hanoi" 
     (trước) và "of Vietnam" (sau) để đưa ra dự đoán chính xác.
   - BERT được huấn luyện trước (pre-trained) với tác vụ Masked Language Modeling (MLM), 
     nên nó rất giỏi trong việc dự đoán từ bị che giấu dựa trên ngữ cảnh xung quanh.
   - Các mô hình Decoder-only (như GPT) chỉ nhìn một chiều (từ trái sang phải) nên 
     không thể tận dụng thông tin từ các từ phía sau token [MASK].

# BÀI 2: DỰ ĐOÁN TỪ TIẾP THEO

In [5]:
from transformers import pipeline

generator = pipeline("text-generation")
prompt = "The best thing about learning NLP is"
generated_texts = generator(prompt, max_length=50, num_return_sequences=3)

print(f"Câu mồi: '{prompt}'\n")
print("=" * 70)
for i, text in enumerate(generated_texts, 1):
    print(f"\nVăn bản {i} được sinh ra:")
    print(text['generated_text'])
    print("-" * 70)

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
d:\10. ky1nam4\NLP\venv\lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in E:\hf_cache\hub\models--openai-community--gpt2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows

Câu mồi: 'The best thing about learning NLP is'


Văn bản 1 được sinh ra:
The best thing about learning NLP is its ability to teach. There is no better time to learn NLP than right now!

A few years ago I wrote about how to practice using NLP as a teaching tool for aspiring programmers. In this post I'll be talking about some of my own experiences with NLP.

The first time I learned NLP was when I was working on a project like this:

So I was working on a very complex and very complex, very big, world, with a very large team. I tried to use NLP with great success, and in the end I got stuck.

I ended up with an NLP project that I managed to pull off so easily, I had to do very little. I mean, it was so easy. I was able to make a simple app that was pretty easy to use.

The next time I found myself in a very complex world, I tried to use it:

But I had to make a few specific mistakes, which I did not make any mistakes on, and also because I was stuck in a very complex world, I did not h

1. Kết quả sinh ra có hợp lý không?

   Kết quả thường khá hợp lý về mặt ngữ pháp và có tính mạch lạc. 
   Tuy nhiên, nội dung có thể không hoàn toàn chính xác về mặt thực tế hoặc 
   có thể hơi chung chung. Mô hình GPT-2 mặc định được huấn luyện trên dữ liệu 
   tiếng Anh nên nó có khả năng sinh văn bản tiếng Anh tự nhiên và có ngữ nghĩa.

2. Tại sao các mô hình Decoder-only như GPT lại phù hợp cho tác vụ này?
    
   - GPT được thiết kế với cơ chế Self-Attention một chiều (unidirectional/causal), 
     chỉ xem xét các token đã xuất hiện trước đó. Điều này phù hợp với tác vụ 
     sinh văn bản tuần tự (từ trái sang phải).
   - GPT được huấn luyện với mục tiêu Next Token Prediction, tức là dự đoán token 
     tiếp theo dựa trên các token trước đó. Đây chính là bản chất của tác vụ text generation.
   - Kiến trúc Decoder-only cho phép mô hình sinh từng token một cách tự hồi quy 
     (autoregressive), phù hợp với việc tạo ra chuỗi văn bản dài và mạch lạc.
   - Các mô hình Encoder-only (như BERT) không phù hợp vì chúng được thiết kế để 
     hiểu ngữ cảnh (understanding), không phải sinh văn bản (generation). BERT nhìn 
     cả hai chiều nên không thể sinh văn bản tuần tự một cách tự nhiên.

# BÀI 3: TÍNH TOÁN VECTOR BIỂU DIỄN CỦA CÂU

In [6]:
import torch
from transformers import AutoTokenizer, AutoModel

model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)

print(f"Đã tải mô hình: {model_name}")
print(f"Số tham số của mô hình: {sum(p.numel() for p in model.parameters()):,}")

sentences = ["This is a sample sentence."]
inputs = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')

print("Thông tin tokenization:")
print(f"Input IDs: {inputs['input_ids']}")
print(f"Attention Mask: {inputs['attention_mask']}")
print(f"Tokens: {tokenizer.convert_ids_to_tokens(inputs['input_ids'][0])}")


with torch.no_grad():
    outputs = model(**inputs)

last_hidden_state = outputs.last_hidden_state
print(f"\nShape của last_hidden_state: {last_hidden_state.shape}")
print(f"  - batch_size: {last_hidden_state.shape[0]}")
print(f"  - sequence_length: {last_hidden_state.shape[1]}")
print(f"  - hidden_size: {last_hidden_state.shape[2]}")

attention_mask = inputs['attention_mask']
mask_expanded = attention_mask.unsqueeze(-1).expand(last_hidden_state.size()).float()
sum_embeddings = torch.sum(last_hidden_state * mask_expanded, 1)
sum_mask = torch.clamp(mask_expanded.sum(1), min=1e-9)
sentence_embedding = sum_embeddings / sum_mask

print("\n" + "=" * 70)
print("VECTOR BIỂU DIỄN CỦA CÂU:")
print("=" * 70)
print(f"Shape: {sentence_embedding.shape}")
print(f"\nVector (10 giá trị đầu tiên): {sentence_embedding[0][:10]}")
print(f"\nNorm của vector: {torch.norm(sentence_embedding).item():.4f}")

Đã tải mô hình: bert-base-uncased
Số tham số của mô hình: 109,482,240
Thông tin tokenization:
Input IDs: tensor([[ 101, 2023, 2003, 1037, 7099, 6251, 1012,  102]])
Attention Mask: tensor([[1, 1, 1, 1, 1, 1, 1, 1]])
Tokens: ['[CLS]', 'this', 'is', 'a', 'sample', 'sentence', '.', '[SEP]']

Shape của last_hidden_state: torch.Size([1, 8, 768])
  - batch_size: 1
  - sequence_length: 8
  - hidden_size: 768

VECTOR BIỂU DIỄN CỦA CÂU:
Shape: torch.Size([1, 768])

Vector (10 giá trị đầu tiên): tensor([-6.3874e-02, -4.2837e-01, -6.6779e-02, -3.8430e-01, -6.5785e-02,
        -2.1826e-01,  4.7636e-01,  4.8659e-01,  3.9991e-05, -7.4274e-02])

Norm của vector: 10.4174


In [7]:
cls_embedding = last_hidden_state[:, 0, :]

print("\n" + "=" * 70)
print("SO SÁNH HAI PHƯƠNG PHÁP:")
print("=" * 70)
print(f"1. Mean Pooling - Shape: {sentence_embedding.shape}")
print(f"   Norm: {torch.norm(sentence_embedding).item():.4f}")
print(f"\n2. CLS Token - Shape: {cls_embedding.shape}")
print(f"   Norm: {torch.norm(cls_embedding).item():.4f}")

cosine_sim = torch.nn.functional.cosine_similarity(
    sentence_embedding, cls_embedding, dim=1
)
print(f"\nCosine Similarity giữa hai phương pháp: {cosine_sim.item():.4f}")


SO SÁNH HAI PHƯƠNG PHÁP:
1. Mean Pooling - Shape: torch.Size([1, 768])
   Norm: 10.4174

2. CLS Token - Shape: torch.Size([1, 768])
   Norm: 14.7415

Cosine Similarity giữa hai phương pháp: 0.4104


1. Kích thước (chiều) của vector biểu diễn là bao nhiêu? Con số này tương ứng với
tham số nào của mô hình BERT?

Con số này tương ứng với tham số hidden_size của mô hình BERT-base. Đây là số chiều của vector embedding mà mỗi token được biểu diễn trong các hidden layers. Từ kết quả last_hidden_state.shape = [1, 8, 768], ta thấy mỗi token trong 8 tokens của câu đều được biểu diễn bằng vector 768 chiều.

2. Tại sao chúng ta cần sử dụng attention_mask khi thực hiện Mean Pooling?

Chúng ta cần attention_mask để loại bỏ padding tokens khỏi phép tính trung bình.
Khi xử lý batch nhiều câu có độ dài khác nhau, tokenizer sẽ đệm (padding) các câu ngắn hơn. Padding tokens không mang ý nghĩa ngữ nghĩa, nếu tính vào trung bình sẽ làm méo mó vector biểu diễn.